service-locator-dns/src/main/scala/com/lightbend/dns/locator/ServiceLocator.scala

```diff
@@ -0,0 +1,154 @@
+
+  /**
+   * A sequence of ServiceAddress objects are the reply, which may of course be empty.
+   */
+  case class Addresses(addresses: Seq[ServiceAddress])
+
+  /**
+   * Used within replies.
+   */
+  case class ServiceAddress(protocol: String, host: String, port: Int)
+
+  private[locator] case class RequestContext(replyTo: ActorRef, resolveOne: Boolean, srv: Seq[SRVRecord])
+
+  private[locator] case class ReplyContext(resolutions: Seq[(Dns.Resolved, SRVRecord)], rc: RequestContext)
+
+  @tailrec
+  private[locator] def matchName(name: String, nameTranslators: Seq[(Regex, String)]): Option[String] =
+    nameTranslators.headOption match {
+      case Some((r, s)) =>
+        val matcher = r.pattern.matcher(name)
+        if (matcher.matches())
+          Some(matcher.replaceAll(s))
+        else
+          matchName(name, nameTranslators.tail)
+      case None => None
+    }
+
+  private[locator] def protocolFromName(name: String): String =
+    name.dropWhile(_ != '.').drop(1).takeWhile(_ != '.').drop(1)
+}
+
+/**
+ * A service locator that can get all addresses for a service using DNS SRV lookups.
+ */
+class ServiceLocator extends Actor with ActorSettings with ActorLogging {
+
+  import ServiceLocator._
+
+  private val _dns = IO(Dns)(context.system)
+  protected def dns: ActorRef = _dns
+
+  private val random = new Random()
+
+  override def receive: Receive = {
+    case GetAddress(name) =>
+      resolve(name, resolveOne = true)
+
+    case GetAddresses(name) =>
+      resolve(name, resolveOne = false)
+
+    case rc: RequestContext =>
+      // TODO: Attend to priority and weight, although they appear pretty useless
+      // When we return just one address then we randomize which of the candidates to return
+      val srvFrom = if (rc.resolveOne && rc.srv.nonEmpty) random.nextInt(rc.srv.size) else 0
+      val srvSize = if (rc.resolveOne && rc.srv.nonEmpty) 1 else rc.srv.size
+      import context.dispatcher
+      val resolutions =
+        rc.srv
+          .slice(srvFrom, srvFrom + srvSize)
+          .map { srv =>
+            dns
+              .ask(Dns.Resolve(srv.target))(settings.resolveTimeout)
+              .mapTo[Dns.Resolved]
+              .map(_ -> srv)
+          }
+      Future
+        .sequence(resolutions)
+        .map(ReplyContext(_, rc))
+        .pipeTo(self)
+
+    case ReplyContext(resolutions, rc) =>
+      val addresses =
+        resolutions
+          .flatMap {
+            case (resolved, srv) =>
+              val protocol = protocolFromName(srv.name)
+              val port = srv.port
+              resolved.ipv4.map(host => ServiceAddress(protocol, host.getHostAddress, port)) ++
+                resolved.ipv6.map(host => ServiceAddress(protocol, host.getHostAddress, port))
+          }
+      rc.replyTo ! Addresses(addresses)
+
+    case iobe: IndexOutOfBoundsException =>
+      log.error("Could not substitute the service name with the name translator {}", iobe.getMessage)
```

Comment:
Let's add a prefix text before the message, e.g. "with the name translator due to: {}"